coreneuron/io/phase1.cpp

```diff
@@ -0,0 +1,170 @@
+
+    F.close();
+}
+
+void Phase1::shift_gids(int imult) {
+    int zz = imult * maxgid;  // offset for each gid
+
+    // offset the (non-negative) gids according to multiple
+    // make sure everything fits into gid space.
+    for (auto& gid: this->output_gids) {
+        if (gid >= 0) {
+            nrn_assert(gid < maxgid);
+            gid += zz;
+        }
+    }
+
+    for (auto& srcgid: this->netcon_srcgids) {
+        if (srcgid >= 0) {
+            nrn_assert(srcgid < maxgid);
+            srcgid += zz;
+        }
+    }
+}
+
+void Phase1::add_extracon(NrnThread& nt, int imult) {
+    if (nrn_setup_extracon <= 0) {
+        return;
+    }
+
+    // very simplistic
+    // Use this threads positive source gids - zz in nt.netcon order as the
+    // source gids for extracon.
+    // The edge cases are:
+    // The 0th duplicate uses uses source gids for the last duplicate.
+    // If there are fewer positive source gids than extracon, then keep
+    // rotating through the nt.netcon .
+    // If there are no positive source gids, use a source gid of -1.
+    // Would not be difficult to modify so that random positive source was
+    // used, and/or random connect to another duplicate.
+    // Note that we increment the nt.n_netcon at the end of this function.
+    int sidoffset = 0;  // how much to increment the corresponding positive gid
+    // like ring connectivity
+    if (imult > 0) {
+        sidoffset = -maxgid;
+    } else if (nrn_setup_multiple > 1) {
+        sidoffset = (nrn_setup_multiple - 1) * maxgid;
+    }
+    // set up the extracon srcgid_
+    int* nc_srcgid = netcon_srcgid[nt.id];
+    int j = 0;  // rotate through the n_netcon netcon_srcgid
+    for (int i = 0; i < nrn_setup_extracon; ++i) {
+        int sid = -1;
+        for (int k = 0; k < nt.n_netcon; ++k) {
+            // potentially rotate j through the entire n_netcon but no further
+            sid = nc_srcgid[j];
+            j = (j + 1) % nt.n_netcon;
+            if (sid >= 0) {
+                break;
+            }
+        }
+        if (sid < 0) {  // only connect to real cells.
+            sid = -1;
+        } else {
+            sid += sidoffset;
+        }
+        nc_srcgid[nt.n_netcon + i] = sid;
+    }
+    // finally increment the n_netcon
+    nt.n_netcon += nrn_setup_extracon;
+}
+
+void Phase1::populate(NrnThread& nt, int imult) {
+    nt.n_presyn = this->output_gids.size();
+    nt.n_netcon = this->netcon_srcgids.size();
+
+    shift_gids(imult);
+
+    netcon_srcgid[nt.id] = new int[nt.n_netcon + nrn_setup_extracon];
+    std::copy(this->netcon_srcgids.begin(), this->netcon_srcgids.end(),
+              netcon_srcgid[nt.id]);
+
+    nt.netcons = new NetCon[nt.n_netcon + nrn_setup_extracon];
+    nt.presyns_helper = (PreSynHelper*)ecalloc_align(nt.n_presyn, sizeof(PreSynHelper));
+
+    nt.presyns = new PreSyn[nt.n_presyn];
+    PreSyn* ps = nt.presyns;
+    for (auto& gid: this->output_gids) {
```

Comment:
```suggestion
    /// go through all presyns
    for (auto& gid: this->output_gids) {
```
just nitpick as a matter of style : with `++ps` at two places in the code, the previous code with explicit loop looks good to my eyes :)

```
for (size_t i = 0; i < nt.n_presyn; ++i) {
   PreSyn* ps = nt.presyns + i;
}
```